File: game-ui-harnesses/ui-component-harness/src/scrollbar-thumb-slices.ts
/** Vertical-only source-pixel slicing; independent of registered track insets. */
export interface ScrollbarThumbSlices {version:'1.0';coordinateSpace:'thumb-source-pixels';top:number;bottom:number}
export function scrollbarThumbSlicesError(value:unknown,height:number,hasInsets:boolean):string|undefined{
 if(!hasInsets)return 'scrollbarInsets required to guarantee a valid minimum thumb size';
 if(!value||typeof value!=='object'||Array.isArray(value))return 'object required';
 const v=value as Record<string,unknown>;
 if(Object.keys(v).sort().join(',')!=='bottom,coordinateSpace,top,version'||v.version!=='1.0'||v.coordinateSpace!=='thumb-source-pixels')return 'unsupported fields, version or coordinate space';
 if(typeof v.top!=='number'||typeof v.bottom!=='number'||!Number.isInteger(v.top)||!Number.isInteger(v.bottom)||v.top<0||v.bottom<0)return 'nonnegative integer source pixels required';
 if(!Number.isFinite(height)||height<=0||v.top+v.bottom>=height)return 'slice caps must leave a positive source middle';
}

File: game-ui-harnesses/ui-decomposition-harness/src/ai_ui_decomposition/scrollbar_thumb_slices.py
"""Consumer scrollbar-thumb-slices-v1. Source pixels never become track units."""
from typing import Literal, TypedDict
import math
from .common import require


class ScrollbarThumbSlices(TypedDict):
    version: Literal['1.0']
    coordinateSpace: Literal['thumb-source-pixels']
    top: int
    bottom: int


def validate_thumb_slices(value, height=None, has_insets=False):
    require(has_insets,'SCROLLBAR_THUMB_SLICES_INSETS_REQUIRED')
    require(isinstance(value,dict) and set(value)=={'version','coordinateSpace','top','bottom'} and
            value['version']=='1.0' and value['coordinateSpace']=='thumb-source-pixels','SCROLLBAR_THUMB_SLICES_FIELDS')
    require(all(type(value[k]) in (int,float) and math.isfinite(value[k]) and value[k]>=0 and value[k]==int(value[k]) for k in ('top','bottom')),'SCROLLBAR_THUMB_SLICES_INTEGER')
    if height is not None:
        require(type(height) in (int,float) and math.isfinite(height) and height>0 and value['top']+value['bottom']<height,'SCROLLBAR_THUMB_SLICES_MIDDLE_REQUIRED')
    return value
